Openshift extension fails to pull images, it creates, when quarkus.container-image.group property is used

### Describe the bug

I have an application, which is being deployed to openshift via `quarkus-openshift` extension. After an update to newer SNAPSHOT (somewhere between 22nd and 24th of December 2023), app fails to start on openshift and there is  "failed to pull image" error in openshift events. This can only be reproduced, when I use `quarkus.container-image.group` property.

### Expected behavior

The application starts

### Actual behavior

The application fails to start, `oc get events` contains this:


```
LAST SEEN   TYPE      REASON              OBJECT                     MESSAGE
63s         Normal    Scheduled           pod/app-1-build            Successfully assigned fd-test/app-1-build to ocp4-14-qtbj7-worker-0-7njbr
61s         Normal    AddedInterface      pod/app-1-build            Add eth0 [EDITED] from openshift-sdn
61s         Normal    Pulled              pod/app-1-build            Container image "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:7207a02d23e70b9b07f01446e5cf2916b807e43b2169e35c33c668302e0842f9" already present on machine
61s         Normal    Created             pod/app-1-build            Created container git-clone
61s         Normal    Started             pod/app-1-build            Started container git-clone
56s         Normal    Pulled              pod/app-1-build            Container image "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:7207a02d23e70b9b07f01446e5cf2916b807e43b2169e35c33c668302e0842f9" already present on machine
56s         Normal    Created             pod/app-1-build            Created container manage-dockerfile
56s         Normal    Started             pod/app-1-build            Started container manage-dockerfile
55s         Normal    Pulled              pod/app-1-build            Container image "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:7207a02d23e70b9b07f01446e5cf2916b807e43b2169e35c33c668302e0842f9" already present on machine
55s         Normal    Created             pod/app-1-build            Created container sti-build
55s         Normal    Started             pod/app-1-build            Started container sti-build
60s         Normal    BuildStarted        build/app-1                Build fd-test/app-1 is now running
41s         Normal    BuildCompleted      build/app-1                Build fd-test/app-1 completed successfully
36s         Normal    Scheduled           pod/app-ff8f5d774-j26ph    Successfully assigned fd-test/app-ff8f5d774-j26ph to ocp4-14-qtbj7-worker-0-7njbr
35s         Normal    AddedInterface      pod/app-ff8f5d774-j26ph    Add eth0 [EDITED] from openshift-sdn
19s         Normal    Pulling             pod/app-ff8f5d774-j26ph    Pulling image "fd-test/app:1.0.0-SNAPSHOT"
18s         Warning   Failed              pod/app-ff8f5d774-j26ph    Failed to pull image "fd-test/app:1.0.0-SNAPSHOT": rpc error: code = Unknown desc = reading manifest 1.0.0-SNAPSHOT in EDITED:5000/fd-test/app: manifest unknown
18s         Warning   Failed              pod/app-ff8f5d774-j26ph    Error: ErrImagePull
4s          Normal    BackOff             pod/app-ff8f5d774-j26ph    Back-off pulling image "fd-test/app:1.0.0-SNAPSHOT"
4s          Warning   Failed              pod/app-ff8f5d774-j26ph    Error: ImagePullBackOff
37s         Normal    SuccessfulCreate    replicaset/app-ff8f5d774   Created pod: app-ff8f5d774-j26ph
37s         Normal    ScalingReplicaSet   deployment/app             Scaled up replica set app-ff8f5d774 to 1
```

### How to Reproduce?

1. `git clone git@github.com:fedinskiy/reproducer.git -b openshift-pull-failure`
2. `cd reproducer`
3. `oc new-project my-project`
4.  `mvn -Popenshift package -Dquarkus.container-image.group=my-project`
5. `oc get events`

if p.4 uses `mvn -Popenshift package` or `mvn -Popenshift package -Dquarkus.container-image.group=my-project -Dquarkus.platform.version=3.6.4`, when deployment is succesfull

### Output of `uname -a` or `ver`

4.18.0-513.9.1.el8_9.x86_64

### Output of `java -version`

11.0.21, vendor: Red Hat, Inc.

### Quarkus version or git rev

3e7134afddbc8dce21ae5eab538ea45d4c967acf

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)

### Additional information

Automatic tests detect, that this error started no earlier, that 22nd of December (the commit was probably merged into main either in late hours of 21st or on 22nd)

I've investigated the same bug. Found strange exception at the core (not sure if it's root case), about handling https (see below). This seems to start by upgrading vert.x core to 4.5.1, in quick glance it seems, that code throwing that, was not present in 4.4.6.


```
2024-01-10 10:37:34,754 ERROR [io.fab.kub.cli.inf.imp.cac.Reflector] (vert.x-eventloop-thread-1) listSyncAndWatch failed for v1/namespaces/ts-dempmwmuij/pods, will stop: java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Scheme: https
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
...
Caused by: java.lang.IllegalArgumentException: Scheme: https
	at io.vertx.core.http.WebSocketConnectOptions.setAbsoluteURI(WebSocketConnectOptions.java:305)
	at io.vertx.core.http.WebSocketConnectOptions.setAbsoluteURI(WebSocketConnectOptions.java:286)
	at io.vertx.core.http.WebSocketConnectOptions.setAbsoluteURI(WebSocketConnectOptions.java:33)
	at io.fabric8.kubernetes.client.vertx.VertxHttpClient.buildWebSocketDirect(VertxHttpClient.java:73)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.buildWebSocketOnce(StandardHttpClient.java:255)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$buildWebSocket$12(StandardHttpClient.java:229)
	at io.fabric8.kubernetes.client.utils.AsyncUtils.retryWithExponentialBackoff(AsyncUtils.java:75)
	at io.fabric8.kubernetes.client.utils.AsyncUtils.retryWithExponentialBackoff(AsyncUtils.java:68)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.retryWithExponentialBackoff(StandardHttpClient.java:163)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.buildWebSocket(StandardHttpClient.java:227)
	at io.fabric8.kubernetes.client.http.StandardWebSocketBuilder.buildAsync(StandardWebSocketBuilder.java:45)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.start(WatchConnectionManager.java:75)
	at io.fabric8.kubernetes.client.dsl.internal.AbstractWatchManager.startWatch(AbstractWatchManager.java:318)
	at io.fabric8.kubernetes.client.dsl.internal.AbstractWatchManager.<init>(AbstractWatchManager.java:146)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.<init>(WatchConnectionManager.java:53)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.submitWatch(BaseOperation.java:661)
	at io.fabric8.kubernetes.client.informers.impl.cache.Reflector.startWatcher(Reflector.java:214)
	at io.fabric8.kubernetes.client.informers.impl.cache.Reflector.lambda$listSyncAndWatch$3(Reflector.java:131)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)


```

I think updated bube client fixes issue mentioned in one of the comments (`Caused by: java.lang.IllegalArgumentException: Scheme: https`) but not the issue mentioned in description.

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/ApplyResolveNamesImagePolicyDecorator.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/ApplyResolveNamesImagePolicyDecorator.java
@@ -1,0 +1,13 @@
+package io.quarkus.kubernetes.deployment;
+
+import io.dekorate.kubernetes.decorator.NamedResourceDecorator;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.api.model.PodTemplateSpecFluent;
+
+public class ApplyResolveNamesImagePolicyDecorator extends NamedResourceDecorator<PodTemplateSpecFluent<?>> {
+
+    @Override
+    public void andThenVisit(PodTemplateSpecFluent<?> podTemplate, ObjectMeta meta) {
+        podTemplate.editOrNewMetadata().addToAnnotations("alpha.image.policy.openshift.io/resolve-names", "*").endMetadata();
+    }
+}

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftProcessor.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftProcessor.java
@@ -236,6 +236,8 @@ public class OpenshiftProcessor {
             result.add(new DecoratorBuildItem(new RemoveOptionalFromConfigMapKeySelectorDecorator()));
         }
 
+        result.add(new DecoratorBuildItem(new ApplyResolveNamesImagePolicyDecorator()));
+
         DeploymentResourceKind deploymentKind = config.getDeploymentResourceKind(capabilities);
         switch (deploymentKind) {
             case Deployment:

--- a/integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/OpenshiftWithDeploymentResourceAndLocalLookupTest.java
+++ b/integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/OpenshiftWithDeploymentResourceAndLocalLookupTest.java
@@ -71,6 +71,9 @@ public class OpenshiftWithDeploymentResourceAndLocalLookupTest {
 
                 assertThat(d.getSpec()).satisfies(deploymentSpec -> {
                     assertThat(deploymentSpec.getTemplate()).satisfies(t -> {
+                        assertThat(t.getMetadata()).satisfies(metadata -> assertThat(metadata.getAnnotations()).contains(
+                                entry("alpha.image.policy.openshift.io/resolve-names", "*")));
+
                         assertThat(t.getMetadata()).satisfies(metadata -> assertThat(metadata.getLabels()).containsAnyOf(
                                 entry("app.kubernetes.io/name", NAME),
                                 entry("app.kubernetes.io/version", "0.1-SNAPSHOT")));

